### PR TITLE
Optimize APT FA transactions - gas charging and transfering

### PR DESCRIPTION
**Pull request by igor-aptos**
## Description

* optimize apt PFS operations by specializing them in primary_fungible_store, and shortcircuiting unnecessary checks (owner/frozen/dispatchable/burned)
* Add feature flags to default to FA for new accounts, and use exclusively FA (i.e. for apt transfers/gas charging). these not yet ready to be enabled on devnet, but we can for some tests.
* add to genesis to create the mapping (i.e. so it's on on devnet/forge). migrate root account to have FA - as nobody will look at it in the explorer, but useful to test e2e
* make concurrent supply default

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/aptos_account.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/aptos_account.move:50]: This should be and just want to be a bit more cautious - can we test sponsored tx where sender's account is not created to make sure not having AptosCoin CoinStore is okay?

File: aptos-move/framework/aptos-framework/sources/coin.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/coin.move:483]: cc @lightmark. This might not make sense here before the end of the migration, as taking BurnRef means that the conversion for migration no longer works

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:410]: Why this change?

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:449]: nit: is_address_balance_at_least -> is_balance_at_least

Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:449]: Why not use is_balance_at_least? Why do we need this friend-only version?

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:75]: Why is this inline? A long inline function used multiple times = bloating up the bytecode

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:69]: nit: < FungibleStore > is not needed

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:150]: Same question here - why is this inline?

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:158]: Doesn't feel like any apt-specific logic should be in PFS? Can this be in aptos_account?

Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:158]: we are specializing APT PFS to be more efficient, so these are all variants of the same functions defined in this file. (so is easier to look and see)

but I can create aptos_primary_fungible_store.move if you prefer to move them out

File: aptos-move/framework/aptos-framework/sources/coin.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/coin.move:483]: conversion doesn't use BurnRef, but uses fungible_asset::burn_internal

This prevents calling get_paired_burn_ref afterwards. I am not sure if we can clone BurnRef somehow, and return it and keep it in pairing as well.

Also I am not sure if caller can have multiple BurnCapability objects, that each needs to be converted.

@lightmark , thoughts?


For APT, we call this only after features::operations_default_to_fa_apt_store() flag has been set, which will not be set until there are no more APT CoinStores

File: aptos-move/framework/aptos-framework/sources/aptos_account.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:50]: I can run all move unit tests with this enabled to confirm. I need to see whether PFS needs to be created - everything works after migration is complete, but coin::transfer doesn't create PFS on demand, so that's probably needed.

note, new_accounts_default_to_fa_apt_store_enabled() flag will not be enabled until ecosystem support is complete.

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:75]: I made not huge private functions inline, to make things faster/cheaper, and was suggested to inline more things on other PRs.
What's the threshold for "function being too long for inline" in your view? here it is like if + 4 function calls, and below one is if + 2 function calls

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (igor-aptos) [inline aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:233]: to reviewers - want to call this out, as this biggest not-purely-optimization change, but a logical change in which invariants are checked.

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (alnoki) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move:75]: > bloating up the bytecode

@movekevin why is this a problem? The Aptos Framework does not face the same package size constraints as dApps, so what is the concern with trading off more bytecode side for less execution gas?

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move]: moves to aptos_primary_fungible_store.move. 

let me know if apt_primary_fungible_store name (i.e. aptos -> apt) makes more sense

File: aptos-move/framework/aptos-framework/sources/aptos_primary_fungible_store.move
Comment (igor-aptos) [inline aptos-move/framework/aptos-framework/sources/aptos_primary_fungible_store.move:54]: to reviewers - want to call this out, as this biggest not-purely-optimization change, but a logical change in which invariants are checked.

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move]: Yeah apt makes more sense

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move]: To save one function call? I'm not sure if it's worth making the code longer

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:450]: removing reading ObjectCore improves performance by more than 10% (i.e. reduces regression by 10%)

we could theoretically have some private function in object to give us Object<T> without checking for T and ObjectCore, but that seems worse to me (as than the invariants stop holding)

Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:450]: Gotcha sounds good then

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move]: store_exists_inline is a single operation - so inling leaves the byte representation the same. If you prefer - I could just not use store_exists in this file and do "exists<>" itself throughout.

**Comment by igor-aptos** [comment]
chatted with @movekevin offline - will measure what inline brings, but if it has measurable perf improvement, there are no concerns of using inlining for these short functions.

File: aptos-move/framework/aptos-framework/sources/primary_fungible_store.move
Comment (alnoki) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move]: > bloat the bytecode

@movekevin it it appropriate to consider increased bytecode size "bloat" if the bytecode size is a function of reducing gas size?

What is the problem with people looking at the framework for an example of gas-optimized code, especially for critical execution paths?

You can always add in a func comment explaining that things are inlined for the sake of performance

Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/primary_fungible_store.move]: https://github.com/aptos-labs/aptos-core/pull/13194#issuecomment-2099438733

and in general - inline is always done for performance, there is no other reasons to inline.

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:450]: I reminds me when I started with those api with address only.

**Review by lightmark — APPROVED** [review]
Object<T> is really annoying sometimes

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:43]: how should you transfer APT efficiently without wanting to create full account?

only aptos_account::transfer and primary_fungible_store::transfer are options, first one creates account, second one is much less performant.

File: aptos-move/framework/aptos-framework/sources/aptos_account.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/aptos_account.move:55]: your coin.move functions, to keep same invariants, fail if target doesn't have neither coin store nor PFS

so this is required, if people use coin module to transfer APT

basically after I create an account, I should be able to receive APT

File: aptos-move/framework/aptos-framework/sources/transaction_fee.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/transaction_fee.move:212]: AptosFABurnCapabilities can be created only after migration is done. 

if AptosFABurnCapabilities exists, other one doesn't, and so this is the only way to charge gas now

File: aptos-move/framework/aptos-framework/sources/transaction_fee.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/transaction_fee.move:220]: this will be slow, but we should convert to AptosFABurnCapabilities after nobody has CoinStore, and then this path will not be executed

File: aptos-move/framework/aptos-framework/sources/coin.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/coin.move:483]: @lightmark does framework have multiple caps? If so - how are we going to convert them to Refs? do we need to actually clone/copy here with some friend function?

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move:483]: We have two mint caps...lemme think about it.

Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/coin.move:483]: I think we cannot generate refs... mint cap should be stored in stake.move and burn is in transaction_fee.move ( transaction_fee is a friend of stake)

Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/coin.move:485]: ok cool, so for this PR we are good (it is burn ref we are touching).
for transforming mint ref, we need to consolidate transaction_fee and stake modules

**Comment by igor-aptos** [comment]
This is fully ready for review - can I get a second review/approval - @movekevin, @davidw, @runtian-zhou ?

File: aptos-move/framework/aptos-framework/sources/coin.move
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/coin.move:909]: after migration is done... I will remove this function I think.

**Comment by igor-aptos** [comment]
Also tested without inlines, and things are 4% slower. definitely worth having them inlined :)

With inline:
```
transaction_type                                 module_working_set  executor      block_size    expected t/s    t/s
no-op                                                             1  VM                 10000           22300  18310
coin-transfer                                                     1  VM                 10000           13100  12275
account-generation                                                1  VM                  9000            9000   9385

```

without inline:
```
transaction_type          module_working_set  executor      block_size    expected t/s    t/s
no-op                                      1  VM                 10000           22300  17645
coin-transfer                              1  VM                 10000           13100  12042
account-generation                         1  VM                  9000            9000   9013
```

**Review by davidw — CHANGES_REQUESTED** [review]
I have no idea why I have been added to this repo.

**Comment by igor-aptos** [comment]
(sorry, typo between davidw and davidiw)

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:13]: styling: Can you move the entry and public functions to the top, followed by public(friend) and with the private functions at the end? This would make it much easier to navigate

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:481]: I don't see why this has to be friend only. Let's just keep it public?

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:34]: Does this have to be public(friend) or can be it public?

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:13]: sure, didn't know that styling guidance

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:39]: nit: < Metadata > is not needed

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:481]: we have public is_balance_at_least<T>(store: Object<T>)

it's not clear for usecases other than gas charging, that it is worth optimization this brings, or that we want to encourage people to not use objects.
keeping things public(friend) means we can also remove them later. public thing is forever

File: aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/apt_primary_fungible_store.move:2]: Might have forgotten some context here - why do we need a standalone module? With the recent dispatchable fungible asset, I think our API/interface is getting a bit confusing with the introduction of many separate but similar sounding/seeming modules

**Review by movekevin — APPROVED** [review]

**Pull request merged**
